webui/react/src/e2e/models/components/Table/InteractiveTable.ts

```diff
@@ -0,0 +1,203 @@
+type RowTypeGeneric<RowType> = new ({ parent, selector }: NamedComponentArgs) => RowType;
+
+export type TableArgs<RowType, HeadRowType> = NamedComponentArgs & {
+  rowType: RowTypeGeneric<RowType>;
+  headRowType: new ({ parent, selector }: NamedComponentArgs) => HeadRowType;
+};
+
+/**
+ * Returns a representation of the InteractiveTable component.
+ * This constructor represents the contents in src/components/Table/InteractiveTable.tsx.
+ * @param {object} obj
+ * @param {implementsGetLocator} obj.parent - The parent used to locate this InteractiveTable
+ * @param {string} [obj.selector] - Used instead of `defaultSelector`
+ */
+export class InteractiveTable<
+  RowType extends Row,
+  HeadRowType extends HeadRow,
+> extends NamedComponent {
+  static defaultSelector = 'div[data-test-component="interactiveTable"]';
+  constructor({ selector, parent, rowType, headRowType }: TableArgs<RowType, HeadRowType>) {
+    super({ parent: parent, selector: selector || InteractiveTable.defaultSelector });
+    this.table = new Table({ headRowType, parent: this, rowType });
+  }
+
+  readonly table: Table<RowType, HeadRowType>;
+  readonly skeleton: SkeletonTable = new SkeletonTable({ parent: this });
+}
+
+/**
+ * Returns the representation of a Table.
+ * This constructor represents the Table in src/components/Table/InteractiveTable.tsx.
+ * @param {object} obj
+ * @param {parentTypes} obj.parent - The parent used to locate this Table
+ * @param {string} obj.selector - Used as a selector uesd to locate this object
+ */
+export class Table<RowType extends Row, HeadRowType extends HeadRow> extends NamedComponent {
+  static defaultSelector = '[data-testid="table"]';
+  constructor({ parent, selector, rowType, headRowType }: TableArgs<RowType, HeadRowType>) {
+    super({ parent: parent, selector: selector || Table.defaultSelector });
+    this.#rowType = rowType;
+    this.rows = new rowType({ parent: this.#body });
+    this.headRow = new headRowType({ parent: this.#head });
+    this.getRowByDataKey = this.rowByAttributeGenerator(this.rows.keyAttribute);
+  }
+  readonly #rowType: RowTypeGeneric<RowType>;
+  readonly rows: RowType;
+  readonly headRow: HeadRowType;
+  readonly #body: BaseComponent = new BaseComponent({
+    parent: this,
+    selector: 'tbody.ant-table-tbody',
+  });
+  readonly #head: BaseComponent = new BaseComponent({
+    parent: this,
+    selector: 'theader.ant-table-thead',
+  });
+  readonly pagination: Pagination = new Pagination({
+    parent: this._parent,
+  });
+
+  /**
+   * Returns a function that gets a row by an attribute value.
+   * @param {string} key - name of the row attribute
+   */
+  rowByAttributeGenerator(key: string): (value: string) => RowType {
+    /**
+     * Returns a row by an attribute value.
+     * @param {string} value - value of the row attribute
+     */
+    return (value: string) => {
+      // TODO default selector should be instance property to make this easier. We want RowType.defaultSelector
```

Comment:
could you talk a little about how making the default selector an instance property would make this easier?

Reply:
that's fixable -- the issue you're running into is that the type of a class is its instance type in typescript -- when referring to the class itself, you want the `typeof` the class:

``` tsx
class Coordinates {
  static memo = "hello" as const
  x = 0
  y = 0
}

const c: Coordinates = new Coordinates();
const s: (typeof Coordinates)['memo'] = "hello"
```

Reply:
yeah I didn't like using the `[]` syntax to access object properties. It was also telling me a few errors when I had tried
<img width="1057" alt="image" src="https://github.com/determined-ai/determined/assets/5224116/f9d775f9-0dc9-4c73-ad88-a56a793ac77d">

Reply:
the error message is saying that you're using the type where you want to use a value -- shouldn't it be `this.rowType.defaultSelector`?

Reply:
yes, but TS doesn't actually know that the static value exists
<img width="858" alt="image" src="https://github.com/determined-ai/determined/assets/5224116/88776d1d-928e-4236-8855-6c872001e7c7">

Reply:
and here's the implementation from basecomponent.ts
```
export abstract class NamedComponent extends BaseComponent {
  constructor({ parent, selector }: { parent: parentTypes; selector: string }) {
    super({ parent, selector });
    requireStaticArgs(this.constructor, ['defaultSelector']);
  }
}

// TODO remove this for instance properties instead. static has been difficult to work with
export function requireStaticArgs<T extends object>(obj: T, requiredProperties: string[]): void {
  requiredProperties.forEach((requiredProp) => {
    if (!Object.hasOwn(obj, requiredProp)) {
      throw new Error(`${obj} must declare a static ${requiredProp}!`);
    }
  });
}
```

Reply:
that's because of the `RowTypeGeneric` -- you don't need that with the `typeof` because it knows the type of the constructor already and changing the type of the constructor would break the extension